HW0/P4/P4.py

```diff
@@ -30,47 +32,114 @@
 
     #####################
     # Part 2:
     #
     # Read the observation array and plot it (Part 2)
     #####################
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
+    
+    ObsTraj = np.loadtxt("P4_measurements.txt",dtype='float',delimiter=',')
+    xObs_coords, yObs_coords, zObs_coords = ObsTraj[:,0], ObsTraj[:,1], ObsTraj[:,2]
+    
+    #Plotting the measured position
+    #ax.plot(xObs_coords, yObs_coords, zObs_coords,'.g', label='Observed trajectory')
+    
+    xtld = ObsTraj.dot(np.diag([1/rx, 1/ry, 1/rz])) # I decided to try and streamline this...
+                                           #  which may be a sign that I'm getting comfy
+    #Plotting the approximate position
+    ax.plot(xtld[:,0],xtld[:,1],xtld[:,2],'.r', label='Approximate Position')                                        
+    
+    
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
+	#Propagation Matrix
+    A = np.diag([1, 1, 1, 1-c*dt, 1-c*dt, 1-c*dt]) + np.diag([dt, dt, dt],3)
+    
+    #Accounting for gravity in update
+    a = np.zeros((A.shape[0])) 
+    a[-1] = g*dt
+    a = a.reshape([6,1])
     # s = ?
 
     # Initial conditions for s0
+    s0 = np.matrix(TruTraj[0,:]).T
+    
     # Compute the rest of sk using Eq (1)
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    sK = np.matrix(np.zeros((6,np.amax(TruTraj.shape)))) #Initializing array for state predictions
+    sK[:,0] = s0                              #Placing initial measurement as first column
+    
+    for i in range(1,np.amax(TruTraj.shape)):
+         sK[:,i] = A * sK[:,i-1] + a
+    
+    sK = np.asarray(sK) # Converting state prediction matrix to an array (for plotting??)
+    
+    #Plotting the predicted positions
+    ax.plot(sK[0,:], sK[1,:], sK[2,:], '-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
-
+    B = np.diag([bx, by, bz, bvx, bvy, bvz])
+    C = np.concatenate( ( np.diag([rx, ry, rz]), np.zeros((3,3)) ), axis=1 )
+    
     # Initial conditions for s0 and Sigma0
+    s0_kalman, Sigma0 = np.matrix(TruTraj[0,:]).T, np.identity(A.shape[0]) * 0.01
+    
     # Compute the rest of sk using Eqs (2), (3), (4), and (5)
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-r', label='Filtered trajectory')
+    
+    def predictS(priorState): # Eq(2)
+        #Will fill out code structure here based on state update from part 3
+        predS = np.add(np.dot(A,priorState),a)
+        return predS
+         
+    def predictSig(priorSig): # Eq(3)
+        #Predict SigmaK update
+        predSig = np.linalg.inv(np.add(np.dot(np.dot(A, priorSig),np.transpose(A)),np.dot(B,np.transpose(B))))
```

Comment:
You could condense the notation a bit if you wanted to: (1) np.transpose(A) is the same as A.T, (2) np.dot(A, B) is the same as A.dot(B)